### PR TITLE
Tkt 41 context

### PR DESCRIPTION
**Pull request by Mikhail-M**
Anyone can find out information about the context through session.getState(row)
It is just a prototype.
I will use stack and other later
@dbarashev  any suggestions?

### CODE REVIEW

File: lib/ace/mode/papeeria_latex_highlight_rules.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+define(function(require, exports, module) {
+"use strict";
+
+var oop = require("../lib/oop");
+var TextHighlightRules = require("./text_highlight_rules").TextHighlightRules;
+
+var PapeeriaLatexHighlightRules = function() {
```
Comment (dbarashev) [inline lib/ace/mode/papeeria_latex_highlight_rules.js]: did you just copy them from latex_highlight_rules? What's the purpose?

Comment (Mikhail-M) [inline lib/ace/mode/papeeria_latex_highlight_rules.js]: No, no.
We have a completely different state machine here.

New file -- to keep the old highlight

Comment (dbarashev) [inline lib/ace/mode/papeeria_latex_highlight_rules.js]: okay, thanks, I will take closer look then :)

Comment (Mikhail-M) [inline lib/ace/mode/papeeria_latex_highlight_rules.js]: Seems, now I have to design great and correct finite state machine.

Comment (dbarashev) [inline lib/ace/mode/papeeria_latex_highlight_rules.js]: sort of.

**Comment by Mikhail-M** [comment]
@dbarashev seems, It looks ready to merge

File: demo/summer-2016.html [hunk 2]
```diff
@@ -242,7 +242,22 @@
...
+        if(editor.getSession().getContext(editor.getCursorPosition().row) == "list") {
+          editor.insert("\n\t\\item ")
+          return true; // CHANGE HERE ! true is when it capture it.
+        } else {
+          // just leave the key.
+          return false; // CHANGE HERE ! false is when I don't capture it.
+        }},
+  readOnly: true
+});
+editor.keyBinding.addKeyboardHandler(keyboardHandler);
```
Comment (dbarashev) [inline demo/summer-2016.html]: why don't you use `editor.commands.addCommand` ?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 3]
```diff
@@ -0,0 +1,131 @@
+define( (require, exports, module) ->
+  'use strict'
+  oop = require('../../lib/oop')
+  TextHighlightRules = require('../../mode/text_highlight_rules').TextHighlightRules
+
+  PapeeriaLatexHighlightRules = ->
+
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: remove this empty line?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    pushState = (destination) -> (currentState, stack) ->
+      if currentState == 'start'
+        stack.push(currentState, destination)
+      else
+        stack.push(destination)
+      destination
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: did you mean to write `return` here? Or just made a typo?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 5]
```diff
@@ -0,0 +1,131 @@
+define( (require, exports, module) ->
+  'use strict'
+  oop = require('../../lib/oop')
+  TextHighlightRules = require('../../mode/text_highlight_rules').TextHighlightRules
+
+  PapeeriaLatexHighlightRules = ->
+
+    pushState = (destination) -> (currentState, stack) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: pls add a few lines of comments. What this function expects, what does it return. I mean, really the types of objects. Is `destination`are string, boolean, number? 

The chain of two functions looks unreadable. I suggest leaving only function name `pushState` and its arguments  on the first line.

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+      if currentState == 'start'
+        stack.push(currentState, destination)
+      else
+        stack.push(destination)
+      destination
+
+    popState = (currentState, stack) ->
+      stack.pop() or 'start'
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: pls always write `return` statement where you mean to return

File: demo/summer-2016.html [hunk 7]
```diff
@@ -242,7 +242,22 @@
...
 
-
+  var HashHandler = require("ace/keyboard/hash_handler").HashHandler;
+  keyboardHandler = new HashHandler();
+  keyboardHandler.addCommand({
+  name: 'item.',
+  bindKey: {win: 'Enter', mac: 'Enter'},
+  exec: function(editor) {
+        if(editor.getSession().getContext(editor.getCursorPosition().row) == "list") {
+          editor.insert("\n\t\\item ")
```
Comment (dbarashev) [inline demo/summer-2016.html]: it now adds `\item` nearly everywhere, e.g. if you hit Enter just after `\begin{document}`

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        token: 'rparen'
+        regex: '[\\])}]'
+      }
+      {
+        token: 'constant.character.escape'
+        regex: '\\\\[^a-zA-Z]?'
+      }
+    ]
+    beginRule = (text = '\\w*', destination = 'start') -> 
+      {
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: add `return`

File: demo/summer-2016.html (hunk 7 above)
Comment (Mikhail-M) [inline demo/summer-2016.html]: no

Comment (dbarashev) [inline demo/summer-2016.html]: quite convincing

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 9]
```diff
@@ -0,0 +1,136 @@
+define((require, exports, module) ->
+  'use strict'
+  oop = require('../../lib/oop')
+  TextHighlightRules = require('../../mode/text_highlight_rules').TextHighlightRules
+
+  PapeeriaLatexHighlightRules = ->
+    ###*
+    * correct environment into stack
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: what's the meaning of stak in this context? Is it a stack of nested semantic constructions, like document=>list=>list=>maths ?

Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: yes

**Comment by Mikhail-M** [comment]
@dbarashev

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 9 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: so, a good comment looks like this

We maintain a stack of nested LaTeX semantic types (e.g. "document", "section", "list") to be able to provide context for autocompletion and other functions. Stack is constructed by the background highlighter; its elements are then propagated to the editor session <describe here who is responsible for that> and become available through getContext method.

This function adds a new type on the stack.

Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: Thank you =)

**Comment by Mikhail-M** [comment]
mb, LGTM?

File: demo/summer-2016.html (hunk 2 above)
Comment (Mikhail-M) [inline demo/summer-2016.html]: It is not a good idea, because using editor.commands.addCommand we defining default action with  Enter
In the results, we will lose the opportunity to move to a new line using Enter

File: demo/summer-2016.html [hunk 10]
```diff
@@ -240,14 +239,35 @@
...
 
+    var HashHandler = require("ace/keyboard/hash_handler").HashHandler; 
+    keyboardHandler = new HashHandler();
+    keyboardHandler.addCommand({
+      name: 'item.',
+      bindKey: {win: 'Enter', mac: 'Enter'},
+      exec: function(editor) {
+        var pos = editor.getCursorPosition()
+        if (editor.getSession().getContext(pos.row) == "list") {
+          var currline = editor.session.getLine(pos.row);
```
Comment (dbarashev) [inline demo/summer-2016.html]: Usually programmers use "cur" prefix (one r) and capitalize the second word: `curLine`

File: demo/summer-2016.html [hunk 11]
```diff
@@ -240,14 +239,35 @@
...
+          if (currline.indexOf("begin") > pos.column) {
```
Comment (dbarashev) [inline demo/summer-2016.html]: why this line?

File: demo/summer-2016.html [hunk 12]
```diff
@@ -240,14 +239,35 @@
...
+            return false;
+          }
+          editor.insert("\n\t\\item ")
+          return true;
+        }
+        else {
```
Comment (dbarashev) [inline demo/summer-2016.html]: you can just remove else

File: demo/summer-2016.html [hunk 13]
```diff
@@ -240,14 +239,35 @@
...
+          return false;
+        }},
+        readOnly: true
```
Comment (dbarashev) [inline demo/summer-2016.html]: what do you mean `readOnly: true` ?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+  'use strict'
+  oop = require('../../lib/oop')
+  TextHighlightRules = require('../../mode/text_highlight_rules').TextHighlightRules
+
+  PapeeriaLatexHighlightRules = ->
+    ###*
+    * We maintain a stack of nested LaTeX semantic types (e.g. "document", "section", "list"
+    * to be able to provide context for autocompletion and other functions. 
+    * Stack is constructed by the background highlighter;
+    * its elements are then propagated to * the editor session and become
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: I thought I asked to explain _how_ are they propagated. It is not quite obvious how `bgTokenizer.getState` becomes aware of all this stuff

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    * We maintain a stack of nested LaTeX semantic types (e.g. "document", "section", "list"
+    * to be able to provide context for autocompletion and other functions. 
+    * Stack is constructed by the background highlighter;
+    * its elements are then propagated to * the editor session and become 
+    * available through getContext method.
+
+    * @param {destination} string
+    * @return {function} function, which correctly puts new type(destination) on stack
+   ###
+    pushState = (destination) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: pls rename this `destination` because it is not a destination at all.

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+      return (currentState, stack) ->
+        if currentState == 'start'
+          stack.push(currentState, destination)
+        else
+          stack.push(destination)
+        return destination
+
+    popState = (currentState, stack) ->
+      return stack.pop() or 'start'
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: any sense to use LaTeX terminology in the state names? E.g. "document" as the root state

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+          'rparen'
+        ]
+        regex: '(\\\\(?:end))({)(' + text + ')(})'
+
+        next: popState
+      }
+
+    @$rules =
+      'start': [
+        beginRule('itemize|enumerate', 'list')
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: Any sense to create constants for the types?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+      ]
+
+
+    for key of @$rules
+      for rule of basicRules
+        @$rules[key].push(basicRules[rule])
+
+    return
+
+  oop.inherits PapeeriaLatexHighlightRules, TextHighlightRules
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: always surround arguments with braces

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 17 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: sure, I'll create it

File: demo/summer-2016.html (hunk 13 above)
Comment (Mikhail-M) [inline demo/summer-2016.html]: I don't know
Seems,  that it is unnecessary

Comment (Mikhail-M) [inline demo/summer-2016.html]: It means, that no one can't change this command in runtime

File: demo/summer-2016.html (hunk 12 above)
Comment (Mikhail-M) [inline demo/summer-2016.html]: No, but I can rewrite if

File: demo/summer-2016.html (hunk 13 above)
Comment (dbarashev) [inline demo/summer-2016.html]: wild guess ;) ? No, it means that this command is supposed to work when editor is read-only.

Comment (Mikhail-M) [inline demo/summer-2016.html]: Definitely not, because this command worked perfectly when editor is not read-only=)

Comment (dbarashev) [inline demo/summer-2016.html]: It says that when editor __is__ read-only, this command should work. It doesn't say anything about editor which __is not__ read-only.

Comment (Mikhail-M) [inline demo/summer-2016.html]: Ok
Perhaps you're right.

File: demo/summer-2016.html [hunk 19]
```diff
@@ -240,14 +239,32 @@
         enableLiveAutocompletion: true
     });
 
+    var HashHandler = require("ace/keyboard/hash_handler").HashHandler; 
+    keyboardHandler = new HashHandler();
+    keyboardHandler.addCommand({
+      name: 'item.',
```
Comment (dbarashev) [inline demo/summer-2016.html]: why such strange  name?

File: demo/summer-2016.html [hunk 20]
```diff
@@ -240,14 +239,32 @@
...
+      bindKey: {win: 'Enter', mac: 'Enter'},
+      exec: function(editor) {
+        var pos = editor.getCursorPosition()
+        var curLine = editor.session.getLine(pos.row);
+
+        if (editor.getSession().getContext(pos.row) == "list" && curLine.indexOf("begin") < pos.column) {
```
Comment (dbarashev) [inline demo/summer-2016.html]: pls add a comment explaining these exercises

File: demo/summer-2016.html [hunk 21]
```diff
@@ -240,14 +239,32 @@
...
+          editor.insert("\n\t\\item ")
+          return true;
+        }
+        else {
+          return false;
+        }}
+        });
```
Comment (dbarashev) [inline demo/summer-2016.html]: nit; this should be indented a little bit less

File: lib/ace/background_tokenizer.js [hunk 22]
```diff
@@ -71,7 +71,7 @@ var BackgroundTokenizer = function(tokenizer, editor) {
         var startLine = currentLine;
         while (self.lines[currentLine])
             currentLine++;
-        
+
```
Comment (dbarashev) [inline lib/ace/background_tokenizer.js]: revert this change?

File: lib/ace/edit_session.js [hunk 23]
```diff
@@ -324,6 +324,19 @@ var EditSession = function(text, mode) {
     };
 
     /**
+     * @param {Number} row
+     * 
+     * Returns context at row.
+     **/
+    this.getContext = function(row) {
```
Comment (dbarashev) [inline lib/ace/edit_session.js]: There is `getState` method above. I think we can move the analysis of the returned state to our own code

File: lib/ace/editor.js [hunk 24]
```diff
@@ -578,6 +578,7 @@ var Editor = function(renderer, session) {
             var pos = self.getCursorPosition();
             var iterator = new TokenIterator(self.session, pos.row, pos.column);
             var token = iterator.getCurrentToken();
+            console.log("token:", token, "mode:", session.getContext(pos.row), "at row", pos.row)
```
Comment (dbarashev) [inline lib/ace/editor.js]: remove debug output?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+      {
+        token: 'rparen'
+        regex: '[\\])}]'
+      }
+      {
+        token: 'constant.character.escape'
+        regex: '\\\\[^a-zA-Z]?'
+      }
+    ]
+    listType = 'itemize|enumerate'
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: it is regex-like, it is not a type. 

Pls USE_ALL_UPPERCASE for the constant names

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+    equationType = 'equation|equation\\*'
+    listState = 'list'
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: You may want to export semantic types, so that other modules could use the constants

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 27]
```diff
@@ -1,11 +1,95 @@
...
-        value: word
-        meta: 'TEX'
-      }
-    ))
-  return
+    context = session.getContext(pos.row)
+    if context == "start"
+      callback(null, listSnippets.concat(equationSnippets.concat(basicSnippets)))
+
+    if context == 'list'
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: E.g. this code will definitely benefit from using a named constant exported by the highlight rules

File: lib/ace/snippets.js [hunk 28]
```diff
@@ -884,6 +884,7 @@ var TabstopManager = function(editor) {
             if (exports.snippetManager && exports.snippetManager.expandWithTab(ed)) {
                 return;
             }
+            console.log(exports.snippetManager.expandWithTab(ed)    )
```
Comment (dbarashev) [inline lib/ace/snippets.js]: remove debug output?

File: lib/ace/edit_session.js (hunk 23 above)
Comment (Mikhail-M) [inline lib/ace/edit_session.js]: mb, but right now it usefull
 suggest you leave this for the future

Comment (dbarashev) [inline lib/ace/edit_session.js]: it will remain useful even if moved outside of the editor code. The issue is that it relies on particular convention about the state type (it may be string but it might be array as well) and semantics. It is our own convention and it is natural for our own conventions to be in our own code.

Comment (Mikhail-M) [inline lib/ace/edit_session.js]: Ok, i will move this into tex_completer

Comment (Mikhail-M) [inline lib/ace/edit_session.js]: I suggest make this in the next pull request 
I  would like to draw everything I did as a good class, which will be initiated in the papeeria_mode.js

Comment (dbarashev) [inline lib/ace/edit_session.js]: okay

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 29]
```diff
@@ -0,0 +1,149 @@
+define((require, exports, module) ->
+  'use strict'
+  oop = require('../../lib/oop')
+  TextHighlightRules = require('../../mode/text_highlight_rules').TextHighlightRules
+  LIST_REGEX = 'itemize|enumerate'
+  EQUATIOТ_REGEX = 'equation|equation\\*'
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: s/EQUATIOT/EQUATION/

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 30]
```diff
@@ -1,11 +1,97 @@
 define (require, exports, module) ->
+  EQUATION_STATE = require('./papeeria_latex_highlight_rules').EQUATION_STATE
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: Pls consider requiring highlight rules only once.

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 31]
```diff
@@ -1,11 +1,97 @@
 define (require, exports, module) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: surround arguments with braces

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 32]
```diff
@@ -1,11 +1,99 @@
-define (require, exports, module) ->
+define( (require, exports, module) ->
+  papeeria_latex_highlight_rules = require('./papeeria_latex_highlight_rules')
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: Pls use camelCase or even CapitalizedWords for module names

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 33]
```diff
@@ -1,5 +1,6 @@
-define (require, exports, module) ->
-  EQUATION_STATE = require('./papeeria_latex_highlight_rules').EQUATION_STATE
+define( (require, exports, module) ->
+  papeeria_latex_highlight_rules = require('./papeeria_latex_highlight_rules')
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: unless I don't see something between the lines, the number of `require`calls didn't change.

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 34]
```diff
@@ -114,25 +114,25 @@ define((require, exports, module) ->
...
+        beginRule(EQUATION_REGEX, EQUATION_STATE)
 
-        endRule(EQUATIOТ_REGEX)
+        endRule(EQUATION_REGEX)
         endRule(LIST_REGEX)
 
       ]
-      EQUATION_STATE: [
-        beginRule(EQUATIOТ_REGEX, EQUATION_STATE)
+      'equation': [
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: why did you decide to replace constant with string literal?

Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: because our backend doesn't support constant in this place.
You can try undo about two commits and see the console

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 35]
```diff
@@ -111,6 +111,9 @@ define((require, exports, module) ->
         next: popState
       }
 
+    # 	For unknown reasons  we can't use constants in block below, because background_tokenizer
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: nit: remove one space char before "For"

**Comment by dbarashev** [comment]
LGTM
